### DISCUSSION TITLE
ReaLearn stopped detecting myJoystick

### DISCUSSION THREAD

**Discussion by Oldorgan** [helgoboss/helgobox#1389] [Q&A]
I had a working project, using a USB Thrustmaster joydtick as a MIDI controller. I used ReaLearn to map the Joystick inputs and buttons to actions in SonicAnomaly SurroundPan 2. This enabled me to manualy move incoming audio around multiple speakers as part of a theater project. It worked like a charmuntil I tried a different joystick. With the second joystick, the Y-axis (mapped to Pitch shift) did not function, so I switched back, and now nothing works. I have updated/uninstalled and re-installed both Reaper and ReaLearn. The Joystick processing script shows midi Events being generated, Clifton joystick inspector script receives all buttons, 2 Axis, and the POV hat, MIDI Logger can receive input, but in ReaLearn? Crickets. What am I overlooking? Howcan I get ReaLearn to see the MIDI events being generated by the joystick processing script? Any help would be greatly appreciated, as the alternate solution (Joy2Key mapping joystick axis to mouse movements) is a lot less flexible.

**Comment by helgoboss** [comment]
This must be a configuration issue. MIDI works as good as always here.

1. Set input correctly in ReaLearn
2. Enable logging in ReaLearn via Menu => Logging => Real control (or something like that). Then generate some MIDI events.
3. Generate some MIDI events

Do they show up?

Can you make a screenshot of ReaLearn's main panel?

**Comment by Oldorgan** [comment]
I followed the steps you gave to the best of my knowledge
1) Set input correctly in ReaLearn
    - MIDI: <FX input>
    -  The Track inout is set to "All MIDI Inputs -> All channels 
    - I tried it set to Channel 1 as well - no effect.
  2) Enable logging in ReaLearn via Menu => Logging => Real control (or something like that). Then generate some MIDI events.
    - I set the logging, but I do not know wherer the log files are storred.
    - In C:\Users\MyUser\AppData\Local\Temp there are various files e.g.
        - StructuredQuery.log
        - mat-debug-23792.log
        - cv_debug.log
        - Not sure which is the correct one.
3) Generate some MIDI events
    - nothing appers to reach ReaLearn
  
I have added some screenshots
<img width="476" alt="MIDI devices" src="https://github.com/user-attachments/assets/41887ce6-b965-429a-8b18-dea98a2a5845" />
<img width="280" alt="Joystick with button press" src="https://github.com/user-attachments/assets/d7121e55-31b7-4a02-a8bc-ffe2e0fe2438" />
<img width="527" alt="ReaLearn Screen" src="https://github.com/user-attachments/assets/cc90b667-8cc0-48e4-a12e-7bf1f6b5600c" />

As well. I did an "Export to clipboard"  as a JSON file.
[quad_box.json](https://github.com/user-attachments/files/18270745/quad_box.json)

**Reply by helgoboss** [reply]
The log messages should appear directly in REAPER, not in a file. That means no messages arrived.

I think I see why: Your track input is "All MIDI Inputs" but in the MIDI preferences, "All" is not checked.

BTW, why not select the MIDI device directly in ReaLearn?

**Comment by Oldorgan** [comment]
When creating a Joystick device (Add virtual MIDI...) there is no option to set to "All". I observed within the "Edit Joystick definition" the MIDI events are being created on Channel 1, so I changed my Track input to channel 1, but still no Midi Input

I put the MIDI Logger Plugin directly after ReaLearn in my FX Chain, and it is able to receive MIDI messages
<img width="254" alt="MIDI Logger Learn setting" src="https://github.com/user-attachments/assets/95a10ca5-caf5-4ecd-abf5-e1dee8a6310f" />
<img width="263" alt="MIDI LOGGER input" src="https://github.com/user-attachments/assets/83896258-321e-42e4-8528-007d0e02a05d" />

There is no option in ReaLearn to select the controller directly - it is not in the list
It is a USB Joystick, using a processing script to create MIDI events
Here is the processing script
[joystick.txt](https://github.com/user-attachments/files/18270806/joystick.txt)

I am going to try check if I get input with a normal MIDI device like a keyboard.

**Reply by helgoboss** [reply]
I know you can't set a joystick as input. But it looks like you convert the joystick stuff to some kind of virtual MIDI device called "T.Flight Stick X" and set that as input of the REAPER track. But you can also set this MIDI device as input in ReaLearn directly (instead of `<FX input>`).

If you absolutely have to use `<FX input>`, read this: By "All" I mean double-clicking the MIDI input device in the REAPER preferences and set 'Enable input from this device in "All MIDI devices"'. Alternatively, just choose the "T.Flight Stick X" input device directly as REAPER track input (not "All MIDI devices").

Does the REAPER track itself receive the MIDI messages? If it doesn't receive them, this is what you need to solve. If REAPER doesn't receive the messages, then ReaLearn also won't receive them.

**Comment by Oldorgan** [comment]
Turns out there was something amiss with my computer. Not sure exactly what, but I fixed it.
I tried a normal MIDI input from my keyboard, through a Focusrite Scarlett. via USB
showed up in MIDI devices, enabled input, device can be chosen directly in Reaper track and in ReaLearn  - Still no MIDI event coming through, so ...  not the Joystick/processing script
Tried the same setup on another, older Laptop, everything worked, joystick as well.
Uninstalled some Allen & Heath and Yamaha mixer control software applications.
Re-installed Reaper, without Re-Wire
Everything works once again. Not sure if it was Re-Wire or the other apps, but something was messing with the proper flow of MIDI events.
I thank you for taking the time to help me, and pointing me in the right direction.
